guardian/riff-raff
# Delete obsolate files in Google Cloud Platform deploys

## What does this change?



The datatech team would like to refine the GCS deployment type previously implemented here: https://github.com/guardian/riff-raff/pull/604 and here: https://github.com/guardian/riff-raff/pull/574. We use this deployment type to update our CODE and PROD instances of [Cloud Composer](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjvta_i3Zv5AhXZgVwKHWgAC8cQFnoECAcQAQ&url=https%3A%2F%2Fcloud.google.com%2Fcomposer&usg=AOvVaw0BvoAUnha-00_IH8bqYzJs). This is the Google Cloud hosted implementation of [Apache Airflow](https://airflow.apache.org/]. 

For each deploy we upload a set of (mostly) python files to a GCS bucket. These files contain the [Dags](https://airflow.apache.org/docs/apache-airflow/stable/concepts/dags.html) defining the data-pipelines in the datalake, along with library code that supports them ( for example, modules to run [DBT](https://www.getdbt.com/) models, or create [Dataproc](https://cloud.google.com/dataproc) clusters and submit jobs to them, for example). These are managed as part of the [ophan datalake](https://github.com/guardian/ophan-data-lake/tree/main/airflow) repo in the `dags` and `guardian` directories. 

What we'd like is to alter the GCP dployment type so that when we remove dags ( or some other type of asset ) from our cloud composer code it can, optionally, be removed from the deploy bucket. This is what this does. In order to make this more generic, change the GCP configuration type to take a list of sub-folders and a list of file types and  remove any matching files which are no longer needed. This should be an optional configuration: for CODE, we'll want to retain the current functionality. 

I've added test cases for the new functionality and refactored the GCS test cases to extract common code and make each test a little more  readable.

I'll squash the commits on merge with all all encompassing commit message

## How to test

I deployed this branch to RiffRaff CODE and then used this Airflow branch https://github.com/guardian/ophan-data-lake/pull/6890 to check everything worked. You can replicate this by first deploying build #7373 to airflow CODE and then #7375. The first creates a couple small dag(.py) files in `dags/dags` and then the second deletes them. The deletions are `gcs://europe-west2-datatech-airfl-fdf1e4b5-bucket/dags/dags/commercial_dfp_impressions_backfill_test.py` and  gcs://europe-west2-datatech-airfl-fdf1e4b5-bucket/dags/dags/commercial_dfp_line_item_test.py `. The logs show these being deleted under the `uploadStaticFiles` deployment, task: `GCSUpload`.  Note: We'll only really want to do this for PROD, but in the above airflow builds the changes to `riffraff.yaml` have the configuration altered for CODE for testing purposes




See above:

You'll probably need access to the: [Code Composer](https://console.cloud.google.com/storage/browser/europe-west2-datatech-airfl-fdf1e4b5-bucket/dags;tab=objects?project=datatech-platform-code&prefix=&forceOnObjectsSortingFiltering=false) Dags  folder(ask datatech).

Deploy this branch to riffaff code and then with that deploy **Ophan datalake PR to come** to `ophan:datalake:airflow` to CODE
confirm that there are no errors and that **Details to Come** files have been removed from the deploy bucket


## How can we measure success?



There should be no increase in errors for deploys to  `ophan:datalake:airflow` 


## Have we considered potential risks?



## Images



## Accessibility
